Smart polling implementation

sankalp1999 opened the pull request:
This PR adds smart polling for boards and notes using ETag-based conditional requests and adaptive intervals. It also improves real-time consistency and test reliability across sessions.

- Introduces `lib/hooks/usePolling.ts` (ETag, adaptive intervals, visibility pause, backoff, concurrency guard) and thin wrappers `useBoardNotesPolling` and `useBoardsListPolling`.
- Adds `lib/etag.ts` helpers and updates notes APIs (`app/api/boards/[id]/notes/route.ts`, `app/api/boards/all-notes/notes/route.ts`) to compute/check ETags before fetching full payloads.
- Integrates polling into `app/boards/[id]/page.tsx` and `app/dashboard/page.tsx`; preserves active edits via new `Note` callbacks `onStartNoteEdit`/`onStopNoteEdit`.
- Adds/refactors e2e specs (`tests/e2e/smart-polling.spec.ts`, `tests/e2e/realtime-sync.spec.ts`) and extracts shared mocks (`tests/fixtures/api-mocks.ts`), replacing fixed sleeps with `expect.poll` for stability.
- Improves error handling and lifecycle cleanup (abort on URL change, stop on navigation; redirect to sign-in on 401 in dashboard).
- No breaking UI changes; performance and correctness improvements in background data refresh.

**Edit: 14th August -> We removed etag, added a timestamp tracking based caching mechanism. Removed smart-polling.spec and api-mocks.ts**

Why polling

- Polling is simpler to maintain and debug than WebSockets and SSE. Works well for this small scale project for now.

- If it's deployed on vercel, then it's serverless. will need some external service like supabase/pusher in that case to support websocket

sankalp1999 commented on components/note.tsx:
```diff
@@ -69,6 +69,8 @@ interface NoteProps {
   showBoardName?: boolean;
   className?: string;
   style?: React.CSSProperties;
+  onStartNoteEdit?: (noteId: string) => void;
```
added two optional callbacks to Note: onStartNoteEdit and onStopNoteEdit. The board page uses them to track the actively edited note and skip replacing it during polling updates, so in‑progress edits aren’t overwritten. Backward‑compatible and doesn’t affect callers that don’t pass these props.

sankalp1999 commented:
@slavingia ready for review.

sync on board, notes and checklist item. 4 second polling time (and adjusts based on user activity)


https://github.com/user-attachments/assets/5b20e714-288e-4625-8291-8d22d04cd281

slavingia commented on app/api/boards/[id]/notes/route.ts:
```diff
@@ -8,75 +8,104 @@ import {
 
-    const user = await db.user.findUnique({
-      where: { id: session.user.id },
+    // ETag incorporates notes and checklist items
```
What's ETag?

sankalp1999 replied:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/ETag

slavingia commented:
Any of these changes can be extracted out to an atomic PR first?

sankalp1999 commented:
There are 2 main things happening in this PR

- adding simple polling + adding hooks of same to UI
- adding 2-3 optimisations in which e-tag based caching takes major line count

there are 600 lines of specs

slavingia commented:
Could you benchmark the etag approach vs not? Feels like it's not worth the complexity vs some very simple polling to get "realtime" feel.

sankalp1999 commented:
- Compare ETag vs simple polling on `/api/boards/[id]/notes`.
- Idle only (no mutations).
- Two phases per run: ETag then Simple against the same board.
- Phase duration 20s; users = 4 and 10.
- Metrics: rps(avg), p50, p95, bytes sent, status codes.

| scenario | users | duration | rps(avg) | p50 (ms) | p95 (ms) | bytesSent | bytesSent (human) | statusCodes |
|---|---:|---:|---:|---:|---:|---:|---|---|
| etag-idle | 4 | 20s | 51.9 | 62 | — | 619,089 | 604.6 KB | 304:1037 |
| simple-idle | 4 | 20s | 63.1 | 57 | — | 3,003,414 | 2.86 MB | 200:1263 |
| etag-idle | 10 | 20s | 66.8 | 136 | 281 | 796,995 | 778.3 KB | 304:1335 |
| simple-idle | 10 | 20s | 68.3 | 134 | 286 | 3,250,726 | 3.10 MB | 200:1367 |

Impact:
- ETag cuts network transfer by ~75–80% at idle with similar p50/p95.
- Simple shows slightly higher RPS in these runs but at ~3–4× bandwidth, increasing egress and client JSON work.

sankalp1999 commented:
You will mainly save on bandwidth. Request volume will be the same. If you don't care about this, we can remove it.
Alternatively, we can lower down polling interval to lets say 2 seconds and we will have roughly same bandwidth [I may paste a benchmark]


To reduce request volume, the following changes are already there. Helps flooding the server.

- Tab visibility: pauses fetching when the tab is hidden; resumes on focus.
- Activity-aware backoff: after 30s of no user activity, it doubles the interval (cap 10s). Active default: 4s for notes, 5s for boards.
- Overlap protection: no concurrent fetches; aborts stale requests.
- Error backoff: exponential retries (1s, 2s, 4s), reducing churn on failures.

sankalp1999 commented:
@slavingia do you want to keep e-tag or should i remove it (will have to change specs also)

sankalp1999 commented:
@slavingia ready for review. removed e-tag, added timestamp based mechanism and also redid specs to test against db.

sync is there for notes, checklist items and boards. skipping for archive page for now since it will be less visited. 


https://github.com/user-attachments/assets/3063a7d6-a8dc-4000-bf4b-33ae5ed92014

sankalp1999 commented:
For reviewer's context, removed etag, added a timestamp tracking based caching mechanism. Removed smart-polling.spec and api-mocks.ts

https://github.com/antiwork/gumboard/pull/340#issuecomment-3185091729

sankalp1999 commented:
resolving conflicts...

sankalp1999 commented:
@ershad ready for review again

sankalp1999 commented:
can we try to merge this soon? looks like there are going to be many merge conflicts

sankalp1999 commented:
@slavingia may i know why this was closed?

slavingia commented:
Not a focus for us atm
